Engine/DataFeeds/WorkScheduling/WeightedWorkScheduler.cs

```diff
@@ -0,0 +1,144 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace QuantConnect.Lean.Engine.DataFeeds.WorkScheduling
+{
+    /// <summary>
+    /// This singleton class will create a thread pool to processes work
+    /// that will be prioritized based on it's weight
+    /// </summary>
+    /// <remarks>The threads in the pool will take ownership of the
+    /// <see cref="WorkItem"/> and not share it with another thread.
+    /// This is required because the data enumerator stack yields, which state
+    /// depends on the thread id</remarks>
+    public class WeightedWorkScheduler
+    {
+        /// <summary>
+        /// This is the size of each work sprint
+        /// </summary>
+        internal const int WorkBatchSize = 50;
```

Comment:
Testing the published benchmarks. Wanted it to be as small as possible since this is part of the latency of the system, but not small enough to decrease performance.